core/block_svg.js

```diff
@@ -655,6 +655,87 @@ Blockly.BlockSvg.prototype.showHelp_ = function() {
 };
 
 /**
+ * Creates a callback function for a click on the "duplicate" context menu
+ * option in Scratch Blocks.  The block is duplicated and attached to the mouse,
+ * which acts as though it were pressed and mid-drag.  Clicking the mouse
+ * releases the new dragging block.
+ * @return {Function} A callback function that duplicates the block and starts a
+ *     drag.
+ * @private
+ */
+Blockly.BlockSvg.prototype.duplicateAndDragCallback_ = function() {
+  var oldBlock = this;
+  return function(e) {
+    // Give the context menu a chance to close.
+    setTimeout(function() {
+      var ws = oldBlock.workspace;
+      var svgRootOld = oldBlock.getSvgRoot();
+      if (!svgRootOld) {
+        throw 'oldBlock is not rendered.';
+      }
+
+      // Create the new block by cloning the block in the flyout (via XML).
+      var xml = Blockly.Xml.blockToDom(oldBlock);
+      // The target workspace would normally resize during domToBlock, which
+      // will lead to weird jumps.  Save it for terminateDrag.
+      ws.setResizesEnabled(false);
+
+      // Using domToBlock instead of domToWorkspace means that the new block
+      // will be placed at position (0, 0) in main workspace units.
+      var block = Blockly.Xml.domToBlock(xml, ws);
+      var svgRootNew = block.getSvgRoot();
+      if (!svgRootNew) {
+        throw 'block is not rendered.';
```

Comment:
Only when the block is not rendered, and I don't know what would cause it to not be rendered at the end of the constructor without an error already having been thrown.  In debugging I would use this to get a stack trace and go from there.

I can make this error be specific to duplication for now, but I'm going to combine this code and the flyout code eventually.

I'd like some help with phrasing here.  I can change it to 'newBlock is not rendered' and rename `block` to `newBlock`, but it still doesn't have a lot of information.